Static swagger

euri10 opened the pull request:
About #110 

it adds an optional `static_directory` to FastAPI, if it's None (which it is by default) the generated html uses the cdn existing at present and the icon also.
If it's set to a path, path is added as a starlette StaticFile, and in that case, FastAPI needs an extra dictionary with js, css and favicon

I'd rather discuss if you're ok with that implementation before doing /redoc :see_no_evil: , some stuff might be moved in `setup()`

~and I'm struggling with those mypy errors, would love some help !~

euri10 committed:
Add websocket to APIRouter

euri10 committed:
Merge remote-tracking branch 'upstream/master'

euri10 committed:
Merge remote-tracking branch 'upstream/master'

euri10 committed:
Swagger with static files

euri10 committed:
Swagger with static files

euri10 committed:
Mypy fixes

euri10 committed:
Mypy fixes

euri10 committed:
Raise clear exception if swagger_static dictionnary is not correctly set
More tests, hopefully coverage increase

euri10 committed:
Raise clear exception if swagger_static dictionnary is not correctly set
More tests, hopefully coverage increase

euri10 committed:
Merge remote-tracking branch 'upstream/master'

tiangolo commented:
First, thanks for your work @euri10 !

I have a couple of concerns:

The compiled JS code is almost 1 MB, and at each new Swagger UI version, we would need to update it. And the changes would be huge, as it is not the source code but the "compiled" (transpiled/minified) version.

I think one way to solve it could be to add a script for building the package, that pulls the lastest JS version from NPM (or any CDN) and adds it to the dist files. But I'm still not completely sure how to make Flit "bundle" it, and then how to add it/serve it in FastAPI, being a file included as part of the `fastapi` package itself.

Another option would be to create an additional optional package, that includes those static JS files, and can be installed and used with FastAPI.

Let's keep this on hold, I'll investigate what options do we have, and what's the best/most clean way to solve this use case.

euri10 commented:
I'm not sure I understand your concern, I surely miss something :)

It's just an option users will have:

- either passing nothing aka `app=FastAPI()` which defaults to the cdn version of the files.
- either pass a `static_directory` kwarg and a dict with the 3 relevant files aka `app = FastAPI(static_directory=static_directory, swagger_static={ "js": swagger_static_js, "css": swagger_static_css, "favicon": swagger_static_icon,},)`

If the css / js changes today, you'll update the 2 links in the docs.py file:

https://github.com/tiangolo/fastapi/blob/08d849d5c588f9cf14a38ede9e9d4d0b29a6d9c2/fastapi/openapi/docs.py#L10
https://github.com/tiangolo/fastapi/blob/08d849d5c588f9cf14a38ede9e9d4d0b29a6d9c2/fastapi/openapi/docs.py#L21

Should that option be implemented you'll change them in the spot they're defined by default, I don't see why you'd need to ship them with the library

euri10 committed:
Merge remote-tracking branch 'upstream/master'

euri10 committed:
Merge branch 'master' of github.com:euri10/fastapi

tiangolo commented:
Ah! Sorry, I was misunderstanding something.

There's an actual copy of the Swagger UI JS bundle in the tests. I was thinking that it was meant to be included as the default static version.

---

As this approach would still require users (developers) to download and configure the static files by hand and then pass additional parameters to the app object, it would be quite some effort to set everything up. So, most of the effort required in the original issue #110 would still be needed.

I think I would prefer to:

* Document how to serve static.
* Document how to send HTML responses.
* Document how to combine the two to override the docs.

---

Then, in the case of #110, to reduce as much as possible the steps to get static local docs, I think it could be done with an external installable Python package that includes those static files and can be connected somehow to the app. But I'm not sure yet about all the details.

wshayes commented:
There is already a python package for swagger ui: https://pypi.org/project/swagger-ui-py

tiangolo commented:
Thanks @wshayes ! Very interesting.

I see it has official support for several specific frameworks. Although it seems maybe some of its utils could be re-used.

It seems to be very close to what I think we could use, the drawback I see is that it's several versions behind. But still, it seems promising.

euri10 committed:
Merge remote-tracking branch 'upstream/master'

euri10 committed:
Merge branch 'master' of github.com:euri10/fastapi

euri10 committed:
Merge remote-tracking branch 'upstream/master'

euri10 committed:
Merge remote-tracking branch 'origin/static_swagger' into static_swagger

euri10 committed:
Using a static_directory doesn't require to pass all
3 files, js css and favicon, those that are not
passed default to their cdn version and a warning is
issued

euri10 committed:
Blacked

euri10 committed:
Fixed tests

euri10 committed:
Lint again

tiangolo commented on fastapi/openapi/docs.py:
```diff
@@ -1,45 +1,40 @@
-    return HTMLResponse(
-        """
+def get_swagger_ui_html(
+    *, openapi_url: str, title: str, swagger_locations: Dict
```
Help me change this to accept a `swagger_js_url`, a `swagger_css_url`, and a `swagger_favicon_url`.

All of them `str`, with a default value of the current version.

That way we can document how to re-use these functions.

tiangolo commented on fastapi/openapi/docs.py:
```diff
@@ -1,45 +1,40 @@
+    return HTMLResponse(html)
 
 
 def get_redoc_html(*, openapi_url: str, title: str) -> HTMLResponse:
```
Help me change this to accept a `redoc_js_url`, and a `redoc_favicon_url`.

Both of them being `str`, with a default value of the current URLs.

tiangolo commented on tests/test_local_swagger.py:
```diff
@@ -0,0 +1,52 @@
+                    assert (
+                        f"Using a static directory and missing {missing} so using default"
+                        in caplog.messages
+                    )
```
In this file we will only need to test the functions. Passing an arbitrary string and checking that it's part of the generated HTML in the response.

tiangolo reviewed (changes requested):
Let's do the following, let's simplify the scope of this PR to making the URLs in the docs functions parameterizable, and testing them.

The only files that need to be included in the PR are `fastapi/openapi/docs.py` and the tests.

Then, in a future PR, we can document how to use them to create custom docs responses, overriding the default URLs.

euri10 committed:
Testing docs generation with js, css and favicon parameters

tiangolo commented:
Good job @euri10 , now it's just removing the extra files in `tests/static` and simplifying the tests.

There's no need to have permutations. Just test that the function without parameters returns the fixed URLs, and that adding parameters, they are included in the response.

Then we're good to go with this. The next step will be documenting how to use it, but that will be later.

euri10 committed:
Removed static files

euri10 committed:
Simplified tests

euri10 commented:
> Good job @euri10 , now it's just removing the extra files in `tests/static` and simplifying the tests.
> 
> There's no need to have permutations. Just test that the function without parameters returns the fixed URLs, and that adding parameters, they are included in the response.
> 
> Then we're good to go with this. The next step will be documenting how to use it, but that will be later.

done
I used the function signature to extract defaults so that tests will survive a modification

tiangolo committed:
:art: Add format small changes to openapi/docs.py

tiangolo committed:
:white_check_mark: Separate tests for customizable swagger and redoc

tiangolo commented:
Thanks @euri10 ! :tada: :rocket: 

BTW, nice job refactoring with f-strings :wink: :cake:

ackerleytng commented:
It seems like while `swagger_js_url`, `swagger_css_url`, and `swagger_favicon_url` are parameterized in `get_swagger_ui_html`, there isn't a way to influence these values from creating a `FastAPI` instance. Should we perhaps pass `self.extra` into the inner function at https://github.com/tiangolo/fastapi/blob/master/fastapi/applications.py#L103 ?

(and similarly for redoc)
